src/algorithms/LaurentPoly.jl

```diff
@@ -0,0 +1,220 @@
+
+const hashp_seed = UInt === UInt64 ? 0xcdaf0e0b5ade239b : 0x5ade239b
+
+function Base.hash(p::LaurentPolyElem, h::UInt)
+   for i in terms_degrees(p)
+      c = coeff(p, i)
+      if !iszero(c)
+         h = hash(i, h)
+         h = hash(c, h)
+      end
+   end
+   hash(hashp_seed, h)
+end
+
+# required implementation
+"""
+    terms_degrees(p::LaurentPolyElem) -> AbstractVector{<:Integer}
+
+> Return a vector containing at least all the degrees of the non-null
+> terms of `p`.
+"""
+function terms_degrees end
+
+# return a terms_degrees vector valid for both polys
+function terms_degrees(p::LaurentPolyElem, q::LaurentPolyElem)
+   mdp = terms_degrees(p)
+   mdq = terms_degrees(q)
+   T = promote_type(eltype(mdp), eltype(mdq))
+   @assert T <: Signed # -one(T) must not wrap around
+   if isempty(mdq)
+      mdp, mdq = mdq, mdp
+   end
+   # if only one is empty, it's now mdp
+   if isempty(mdp) # extrema requires non-empty
+      if isempty(mdq)
+         zero(T):-one(T)
+      else
+         UnitRange{T}(extrema(mdq)...) # mdq not returned for type-stability
+      end
+   else # none is empty
+      minp, maxp = extrema(mdp)
+      minq, maxq = extrema(mdq)
+      min(minp, minq):max(maxp, maxq)
+   end
+end
+
+# like terms_degrees, but return a "strict" range, whose min/max
+# corresponds to non-zero coeffs, or an empty range for p == 0
+function degrees_range(p::LaurentPolyElem)
+   mds = terms_degrees(p)
+   T = eltype(mds)
+   minp, maxp = isempty(mds) ? (T(0), T(0)) : extrema(mds)
+   while iszero(coeff(p, minp))
+      minp += 1
+      minp > maxp && return minp:maxp
+   end
+   while iszero(coeff(p, maxp))
+      maxp -= 1
+   end
+   minp:maxp
+end
+
+
+# return the degree of the unique term, throw if not a term
+function term_degree(p::LaurentPolyElem)
+   isnull = true
+   local deg
+   mds = terms_degrees(p)
+   isempty(mds) && throw(DomainError(p, "not a term"))
+   for d in mds
+      if !iszero(coeff(p, d))
+         !isnull && throw(DomainError(p, "not a term"))
+         deg = d
+         isnull = false
+      end
+   end
+   deg
+end
+
+function lead(p::LaurentPolyElem)
+   dr = degrees_range(p)
+   isempty(dr) ? zero(base_ring(p)) : coeff(p, last(dr))
+end
+
+function trail(p::LaurentPolyElem)
+   dr = degrees_range(p)
+   isempty(dr) ? zero(base_ring(p)) : coeff(p, first(dr))
+end
+
+gens(R::LaurentPolynomialRing) = [gen(R)]
+
+isgen(p::LaurentPolyElem) = p == gen(parent(p))
+
+# whether p is a (monic) monomial of degree i, non-recursively:
+# return true iff `p` has only one non-null coefficient
+# (of degree i) at the outer layer, and this coefficient is one
+# (as an element of the base ring)
+function ismonomial_nonrec(p::LaurentPolyElem, i::Integer)
+   dr = degrees_range(p)
+   length(dr) == 1 || return false
+   dr[] == i || return false
+   isone(coeff(p, i))
+end
+
+function ismonomial(p::LaurentPolyElem)
+   dr = degrees_range(p)
+   length(dr) == 1 || return false
+   ismonomial(coeff(p, dr[]))
+end
```

Comment:
Sorry, I think I was not clear enough.

I meant to say the the non-recursive function should be the default. Otherwise this is too confusing and non-standard. If f is a (Laurent polynomial) over a ring R with variable x, then it is a monomial if and only if f = x^i for some i.

Reply:
I would definitely agree, but AFAICT this is not the current definition for polynomials:
```julia
julia> R, x = PolynomialRing(ZZ, "x")
(Univariate Polynomial Ring in x over Integers, x)

julia> S, y = PolynomialRing(R, "y")
(Univariate Polynomial Ring in y over Univariate Polynomial Ring in x over Integers, y)

julia> ismonomial(x*y)
true
```

Reply:
My bad. I was looking at the multivariate polynomials, where it is not recursive. I will open an issue tomorrow. We can keep whatever you did here and adjust it later once we have figured out what we want.

Reply:
Oh right, I didn't notice the discrepancy here with multiviariate polynomials. `isterm` also has the same problem, and its docstring (for the non-multivariate case) explicitly mentions recursivity:

> `isterm(a::Generic.PolynomialElem)`
> Return true if the given polynomial has one term. This function is recursive, with all scalar types returning true.